subsys/bluetooth/mesh/dfw.c

```diff
@@ -0,0 +1,5398 @@
+}
+
+void bt_mesh_dfw_path_origin_state_machine_msg_sent(uint16_t net_idx, uint16_t dst)
+{
+	struct dfw_state_machine *machine = dfw_state_machine_find_by_dst(dst);
+
+	if (!machine) {
+		return;
+	}
+
+	LOG_INF("State machine for dst 0x%04x message sent", dst);
+
+	machine->sent = true;
+}
+
+int bt_mesh_dfw_path_origin_state_machine_start(uint16_t net_idx,
+						const struct bt_mesh_dfw_node *dependent,
+						uint16_t dst, bool power_up)
+{
+	enum bt_mesh_feat_state state = BT_MESH_FEATURE_DISABLED;
+	struct dfw_state_machine *machine;
+	int err;
+
+	err = bt_mesh_dfw_get(net_idx, &state);
+	if (err) {
+		return err;
+	}
+
+	if (state != BT_MESH_FEATURE_ENABLED) {
+		LOG_ERR("Directed forwarding feature not enabled");
+		return -ENOTSUP;
+	}
+
+	if (!dfw_forwarding_dest_addr_is_valid(dst)) {
+		LOG_ERR("State Machine dst address not valid 0x%04x", dst);
+		return -EINVAL;
+	}
+
+	machine = dfw_state_machine_find_by_dst(dst);
+	if (machine) {
+		LOG_ERR("State Machine instance already existed for 0x%04x", dst);
+		return -EALREADY;
+	}
+
+	machine = dfw_state_machine_find(DFW_STATE_MACHINE_FINAL, BT_MESH_ADDR_UNASSIGNED);
+	if (!machine) {
+		LOG_ERR("Insuffcient state machine instance for 0x%04x", dst);
+		return -ENOBUFS;
+	}
+
+	machine->dst    = dst;
+	machine->net_idx = net_idx;
+	machine->fw      = NULL;
+	machine->sent    = false;
+
+	if (dependent) {
+		machine->dependent.addr = dependent->addr;
+		machine->dependent.secondary_count = dependent->secondary_count;
+	}
+
+	if (power_up) {
+		dfw_state_machine_state_set(machine, DFW_STATE_MACHINE_POWER_UP);
+	} else {
+		dfw_state_machine_state_set(machine, DFW_STATE_MACHINE_INITIAL);
+	}
+
+	return 0;
+}
+
+static void dfw_node_encode(const struct bt_mesh_dfw_node *node,
+			    struct net_buf_simple *buf)
+{
+	if (node->secondary_count) {
+		net_buf_simple_add_le16(buf, ((node->addr << 1) | BIT(0)));
+		net_buf_simple_add_u8(buf, node->secondary_count + 1);
+	} else {
+		net_buf_simple_add_le16(buf, node->addr << 1);
+	}
+}
+
+static void dfw_node_decode(struct net_buf_simple *buf,
+			    struct bt_mesh_dfw_node *node)
+{
+	node->addr = net_buf_simple_pull_le16(buf);
+	if (node->addr & BIT(0)) {
+		node->secondary_count = net_buf_simple_pull_u8(buf) - 1;
+	} else {
+		node->secondary_count = 0;
+	}
+
+	node->addr >>= 1;
+}
+
+static int send_ctl_status(const struct bt_mesh_model *model,
+			   struct bt_mesh_msg_ctx *ctx,
+			   struct bt_mesh_subnet *sub,
+			   uint8_t status,
+			   struct net_buf_simple *buf)
+{
+	BT_MESH_MODEL_BUF_DEFINE(msg, OP_DIRECTED_CTL_STATUS, 8);
+	enum bt_mesh_feat_state state = BT_MESH_FEATURE_DISABLED;
+	int err;
+
+	bt_mesh_model_msg_init(&msg, OP_DIRECTED_CTL_STATUS);
+
+	net_buf_simple_add_u8(&msg, status);
+
+	net_buf_simple_add_le16(&msg, net_buf_simple_pull_le16(buf));
+
+	if (status == STATUS_SUCCESS) {
+		(void)bt_mesh_dfw_get(sub->net_idx, &state);
+		net_buf_simple_add_u8(&msg, state);
+
+		(void)bt_mesh_dfw_relay_get(sub->net_idx, &state);
```

Comment:
Why do all such kind of functions in dfw have returned values? It is not analyzed and sent in the air `as is`.